[WGSL] Factor out context logic from MangleNames pass

tadeuzagallo opened the pull request:
#### 0d49272caea1d1837813cb67302e346ffe067936
<pre>
[WGSL] Factor out context logic from MangleNames pass
<a href="https://bugs.webkit.org/show_bug.cgi?id=251523">https://bugs.webkit.org/show_bug.cgi?id=251523</a>
&lt;rdar://problem/104921139&gt;

Reviewed by Myles C. Maxfield.

Move the context logic out of the MangleNames pass so it can be
reused in the type checking pass.

* Source/WebGPU/WGSL/ContextProvider.h: Added.
* Source/WebGPU/WGSL/ContextProviderInlines.h: Added.
(WGSL::ContextProvider&lt;Value&gt;::Context::Context):
(WGSL::ContextProvider&lt;Value&gt;::Context::lookup const):
(WGSL::ContextProvider&lt;Value&gt;::Context::add):
(WGSL::ContextProvider&lt;Value&gt;::ContextScope::ContextScope):
(WGSL::ContextProvider&lt;Value&gt;::ContextScope::~ContextScope):
(WGSL::ContextProvider&lt;Value&gt;::ContextProvider):
(WGSL::ContextProvider&lt;Value&gt;::introduceVariable const):
(WGSL::ContextProvider&lt;Value&gt;::readVariable const const):
* Source/WebGPU/WGSL/MangleNames.cpp:
(WGSL::NameManglerVisitor::run):
(WGSL::NameManglerVisitor::introduceVariable):
(WGSL::NameManglerVisitor::readVariable const):
(WGSL::NameManglerVisitor::Context::Context): Deleted.
(WGSL::NameManglerVisitor::Context::lookup const): Deleted.
(WGSL::NameManglerVisitor::Context::add): Deleted.
(WGSL::NameManglerVisitor::ContextScope::ContextScope): Deleted.
(WGSL::NameManglerVisitor::ContextScope::~ContextScope): Deleted.
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/259748@main">https://commits.webkit.org/259748@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/668d6f6b4ce22cb2e2270f24d31f458a58eda6cf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105693 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14784 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38566 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114918 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16190 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5981 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97957 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114724 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111446 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12315 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95297 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39796 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94179 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26937 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81513 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8049 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28290 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8545 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4875 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14166 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47837 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6741 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10098 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebGPU/WGSL/ContextProviderInlines.h:
```diff
@@ -0,0 +1,94 @@
+}
+
+template<typename Value>
+auto ContextProvider<Value>::introduceVariable(const AST::Identifier& name, Value&& value) -> const Value& 
+{
```
The style checker wants me to move this bracket to the end of the previous line, but I don't think that's correct.

litherum commented on Source/WebGPU/WGSL/ContextProvider.h:
```diff
@@ -0,0 +1,77 @@
+}
+
+template<typename ContextValue>
+class ContextProvider {
```
This sure is a generic name. Can we make it more clear that it's for the mangling pass?

tadeuzagallo replied:
The whole point is that it's not for the mangling pass though, it's for any pass. I guess the name "context" is fairly generic, but not sure what else to name it though. My thinking was: it provides a context in which to look up variables for any pass that needs to keep track of that (e.g. name mangling, type checking, global variable tracking, etc). I believe this often called the context (or environment) in the literature. The only thing that comes to mind is `VariableContextProvider`, but I think that's even worse

litherum replied:
Right, I realized from the second patch that it's used in the type checker too. So I guess it's fine

tadeuzagallo committed:
[WGSL] Factor out context logic from MangleNames pass
https://bugs.webkit.org/show_bug.cgi?id=251523
<rdar://problem/104921139>

Reviewed by Myles C. Maxfield.

Move the context logic out of the MangleNames pass so it can be
reused in the type checking pass.

* Source/WebGPU/WGSL/ContextProvider.h: Added.
* Source/WebGPU/WGSL/ContextProviderInlines.h: Added.
(WGSL::ContextProvider<Value>::Context::Context):
(WGSL::ContextProvider<Value>::Context::lookup const):
(WGSL::ContextProvider<Value>::Context::add):
(WGSL::ContextProvider<Value>::ContextScope::ContextScope):
(WGSL::ContextProvider<Value>::ContextScope::~ContextScope):
(WGSL::ContextProvider<Value>::ContextProvider):
(WGSL::ContextProvider<Value>::introduceVariable const):
(WGSL::ContextProvider<Value>::readVariable const const):
* Source/WebGPU/WGSL/MangleNames.cpp:
(WGSL::NameManglerVisitor::run):
(WGSL::NameManglerVisitor::introduceVariable):
(WGSL::NameManglerVisitor::readVariable const):
(WGSL::NameManglerVisitor::Context::Context): Deleted.
(WGSL::NameManglerVisitor::Context::lookup const): Deleted.
(WGSL::NameManglerVisitor::Context::add): Deleted.
(WGSL::NameManglerVisitor::ContextScope::ContextScope): Deleted.
(WGSL::NameManglerVisitor::ContextScope::~ContextScope): Deleted.
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/259748@main
